README
The Apache Flex SDK
===================

The Apache Flex SDK is the evolution of the popular Adobe Flex SDK. The Apache
Flex SDK is an application development framework for easily building Flash-based
applications for mobile devices, web browsers, and desktop platforms.
Currently supported platforms include:

    Microsoft Windows
    Mac OS X
    Apple iOS
    Google Android
    RIM BlackBerry
    (Linux)

Linux support is currently experimental and you may run into issues. The Apache
Flex project would like to see a fully supported Linux release, but we need
help from the community in order to do that. Please help out if you can.

For more information about the Apache Flex project, please visit:

    http://flex.apache.org/

The Apache Flex SDK 4.16 is a follow up release to version 4.15. It is
compatible with most code written for the original Adobe Flex 4.6 SDK.



Installing the Apache Flex SDK
==============================

The Apache Flex SDK Installer is an AIR application that simplifies the 
download and installation of the Apache Flex SDK and its (required) components.
It is aimed at anyone who wants to use the latest release of the Apache Flex
SDK, but who might not necessarily be familiar with the tools and procedures
required to compile the Apache Flex SDK from source code. The application will
grab the binary distribution of the SDK from apache.org or one of its mirrors,
install it onto your computer and prepare it for use with your favourite IDE
such as Adobe Flash Builder or JetBrains IntelliJ IDEA.

You can get the Apache Flex SDK Installer from the Apache Flex website at:

    http://flex.apache.org/installer.html

Starting with the Apache Flex SDK 4.12, Linux users can install the binary
distribution using ANT. More information about installing the binary package
with ANT can be found on our wiki at:

    https://cwiki.apache.org/confluence/display/FLEX/Installation+help#Installationhelp-Antinstaller



Building the Apache Flex SDK from source
========================================

The Apache Flex SDK is a large project. It requires build tools which must be
installed on your system and it depends on some external software. The external
software is downloaded as part of the build process.

There are two ways to get the SDK source.
1) from a git repository
2) from a source package distribution

The following steps are all you need for a fully functional SDK:
- Prepare your system
    - install external software
    - set environment variables
- Get the source code
- Build the SDK
- Prepare the SDK for use with an IDE
- Optional steps:
  - Flash Player configuration
  - Other locale support


Prepare your system
-------------------

Before building the Apache Flex SDK you must install the following software and
set the corresponding environment variables. Make sure to use absolute paths,
as relative paths will result in build errors.

The environment variables PLAYERGLOBAL_HOME, AIR_HOME, FLASHPLAYER_DEBUGGER,
TLF_HOME, BLAZEDS_HOME and ADOBE_EXTENSION_MANAGER can also be set in a
property file called env.properties. See the env-template.properties file
for instructions.

The Adobe Flash Player and Adobe AIR have non Apache compatible licenses so
they can not be bundled with the Apache Flex SDK and need to be installed
separately.

The Adobe Flash Player ‘playerglobal.swc’ is needed to compile all the
components with the exception of the ‘airframework’ and ‘airspark’
components, which require ‘airglobal.swc’ from the AIR Integration Kit. The
AIR Integration kit is also needed to build and debug mobile and desktop
applications. The Adobe Flash Player content debugger is used by the ant
‘checkintests’ target and other pieces of the test subsystem to run compiled
applications.


---------------------------------------------------------------------
SOFTWARE                                      ENVIRONMENT VARIABLE
---------------------------------------------------------------------

Java SDK 1.6 or greater (* 1)                 JAVA_HOME
    (for Java 1.7 see note at (* 2))

Ant 1.7.1 or greater (* 1)                    ANT_HOME
    (for Java 1.7 see note at (* 2))

Adobe AIR Integration Kit (* 3)               AIR_HOME

Adobe Flash Player Content Debugger (* 4)     FLASHPLAYER_DEBUGGER

Adobe Flash Player playerglobal swcs (* 5)    PLAYERGLOBAL_HOME

Optional: Text Layout Framework (* 6)         TLF_HOME

Optional: Adobe Extension Manager CS5 (* 7)   ADOBE_EXTENSION_MANAGER

Optional: Flex Messaging Common (* 8)         BLAZEDS_HOME

Java Cryptography Extensions
    (see note (*9))

---------------------------------------------------------------------

* 1)
The bin directories for ANT_HOME and JAVA_HOME should be added to your PATH.

On Windows, set PATH to

    PATH=%PATH%;%ANT_HOME%\bin;%JAVA_HOME%\bin

On the Mac (bash), set PATH to

    export PATH='$PATH:$ANT_HOME/bin:$JAVA_HOME/bin'

On Linux make sure your path includes ANT_HOME and JAVA_HOME.

* 2)
If you are using Java SDK 1.7 or greater on a Mac you must use Ant 1.8 or
greater. If you use Java 1.7 with Ant 1.7, ant reports the java version as
1.6 so the JVM args for the data model (-d32/-d64) will not be set correctly
and you will get compile errors.

* 3)
The Adobe AIR integration kit for Windows can be downloaded from:

   http://airdownload.adobe.com/air/win/download/20.0/AdobeAIRSDK.zip

The Adobe AIR integration kit for Mac can be downloaded from:

    http://airdownload.adobe.com/air/mac/download/20.0/AdobeAIRSDK.tbz2

The Adobe AIR integration kit for Linux can be downloaded from:

    http://airdownload.adobe.com/air/lin/download/2.6/AdobeAIRSDK.tbz2

Download the AIR SDK for your platform and unzip it. Set AIR_HOME to the
absolute path of the AIR SDK directory.

This version of the Apache Flex SDK was certified for use with Adobe AIR 16
and is compatible with version 3.1 and up. It has been fully tested on AIR 3.7,
4 and 16.

* 4)
The Adobe Flash Player content debuggers can be found here:

    http://www.adobe.com/support/flashplayer/downloads.html

This version of the Apache Flex SDK was certified for use with Adobe Flash
Player 11.1, and is compatible with version 10.2 and up. It has been tested
with versions 11.1, 11.7, 13.0 and 16.0 on Windows and Mac. It has been
compiled, but not fully tested, with other Adobe Flash Player versions.
It has not been fully tested on Linux.

On Windows, set FLASHPLAYER_DEBUGGER to the absolute path including the
filename of the FlashPlayerDebugger.exe. Note the filename of flash player
debugger may be different, e.g. C:\MyPath\FlashPlayerDebugger.exe.

On the Mac, set FLASHPLAYER_DEBUGGER to the absolute path of
'Flash Player Debugger.app/Contents/MacOS/Flash Player Debugger'

On Linux, set FLASHPLAYER_DEBUGGER to the absolute path of flashplayerdebugger.

* 5)
The Adobe Flash Player playerglobal.swc for 11.1 can be downloaded from:

    http://download.macromedia.com/get/flashplayer/updaters/11/playerglobal11_1.swc

Set PLAYERGLOBAL_HOME to the absolute path of the player directory (not
including the version subdirectory). The target-player option controls which
PLAYERGLOBAL_HOME subdirectory is used.

Additional information about downloading and setting up this and/or other
versions of the Adobe Flash Player can be found in the section
'Adobe Flash Player Version Support' below.

* 6)
The build scripts assume that the root folder containing the source code for
Apache Flex Text Layout Framework ('flex-tlf') is at the same level as the SDK
root folder ('flex-sdk'). See 'Getting the source code', below. If this is not
the case on your system, then you must set the TLF_HOME environment variable to
point to your TLF root folder. If you are using a released artifact, the
source should be included in the artifact, so setting TLF_HOME is not
required.

* 7)
The Adobe Extension Manager is only required for those creating releases or
testing changes to the flash-integration SWC. Note that if you change APIs on
classes that flash-integration depends on, you may need to update
flash-integration.

On Windows, download the Adobe Extension Manager from:

    http://download.macromedia.com/pub/dw_exchange/extension_manager/win/AdobeExtensionManager5All.zip
	
On Windows 10, download the Adobe Extension Manager CS6 from:

	http://download.macromedia.com/pub/dw_exchange/extension_manager/win/AdobeExtensionManager6_0_8All.zip

and set ADOBE_EXTENSION_MANAGER to the absolute path of
'Adobe Extension Manager CS5.exe'

On Mac, download the Adobe Extension Manager from:

    http://download.macromedia.com/pub/dw_exchange/extension_manager/mac/AdobeExtensionManager5All.dmg

and set ADOBE_EXTENSION_MANAGER to the absolute path of
'Adobe Extension Manager CS5.app/Contents/MacOS/Adobe Extension Manager CS5'

On Linux, no Adobe Extension Manager exists.

* 8)
The build scripts assume that the root folder containing the source code for Apache
Flex BlazeDS ('flex-blazeds') is at the same level as the SDK root folder
('flex-sdk'). See 'Getting the source code', below. If this is not
the case on your system, then you must set the BLAZEDS_HOME environment variable
to point to your BLAZEDS root folder. If you are using a released artifact, the
source should be included in the artifact, so setting BLAZEDS_HOME is not
required.

* 9)
The Java Cryptography Extensions (JCE) Unlimited Strength jars are needed in
order to download 3rd party software. Download and install this from Oracle
which version you need will depend on your Java version and OS.

Get the source code
-------------------

You can get the source directly from the source code repository (Apache Flex
uses git for SCM).

Clone the latest source ('develop' branch) using the following commands:

    git clone https://git-wip-us.apache.org/repos/asf/flex-sdk.git flex-sdk
    cd flex-sdk
    git checkout develop

The Apache Flex SDK also requires source code from the Apache Flex Text Layout
Framework (TLF). To get the latest source for TLF ('develop' branch), use the
following commands:

    git clone https://git-wip-us.apache.org/repos/asf/flex-tlf.git flex-tlf
    cd flex-tlf
    git checkout develop

The Apache Flex SDK also requires source code from Apache Flex BlazeDS. To get 
the latest source for Apache Flex BlazeDS (‘master’ branch), use the
following commands:

    git clone https://git-wip-us.apache.org/repos/asf/flex-blazeds.git flex-blazeds
    cd flex-blazeds

Another option to get the source is to download a released source package.

For further information visit:

    http://flex.apache.org/download-source.html


Build the SDK
-------------

When you have all the prerequisites in place and the environment variables set,
use these commands:

    cd <flex-sdk>
    ant main

to download the third party dependencies and build the source. You will be
prompted to acknowledge and/or confirm some of the downloads. Since the
third-party dependencies take a little while to download and they don't change
very often, they are not cleaned with the regular 'clean' target.

If you would like to build the RSLs, use:

    ant frameworks-rsls

To clean the build of everything other than the downloaded third-party
dependencies, use:

    ant clean

To clean the build of everything, including the downloaded third-party
dependencies, use:

    ant super-clean

To generate a source distribution package and a binary distribution package,
use:

    ant -Dbuild.number=<YYYYMMDD> -Dbuild.noprompt= release

The packages can then be found in the <flex-sdk>\out directory.

To build the ASDoc package, use:

    ant asdoc-package

To get a brief listing of all the targets, use:

    ant -projecthelp



Prepare the SDK for use with an IDE
-----------------------------------

To create an SDK for Flash Builder or any other IDE, run:

    cd <flex.dir>
    ant frameworks-rsls
    ant -f installer.xml
    
If you want to set the Flash Player and AIR versions to something other
than their defaults, pass the versions to the ant script like so:

     ant -f installer.xml -Dflash.sdk.version=16.0 -Dair.sdk.version=16.0

On Mac/Linux the AIR and Flash Player version can be changed, after the SDK
has been compiled, by running:

    ./ide/setFlashPlayerVersion.sh ./ 20.0
    ./ide/addAIRtoSDK.sh 20.0 ./



Using the Binary Distribution
=============================

In order to use the binary distribution of the SDK, follow the instructions in
the 'Prepare the SDK for use with an IDE' section above.



Flash Player Configuration
=========================

For testing, the Adobe Flash Player's mm.cfg file must have the following
entries

    ErrorReportingEnable=1
    TraceOutputFileEnable=1

and a FlashPlayerTrust file must allow local SWFs to access local files.

More information about these settings can be found here:

    http://help.adobe.com/en_US/ActionScript/3.0_ProgrammingAS3/WS5b3ccc516d4fbf351e63e3d118a9b90204-7c95.html



Other Locale Support
====================

The Apache Flex SDK defaults to using the en_US locale and SDK error messages
are by default displayed in American English.

To compile the SDK for another locale either:

Change the locale in the build.properties file to have a value other than
'en_US'. For this new locale to take effect, the SDK needs to be recompiled. For
example, to use the Australian English locale change locale to have a value of
'en_AU' like so:

    locale = en_AU

OR

Compile the SDK with a -Dlocale=<locale> option, where <locale> is the new
locale to compile the SDK against.

To compile all supported locales, which may take some time, use:

    ant other.locales

The list of supported locales currently include:
    da_DK    el_GR    en_GB    fi_FI    it_IT    nb_NO    pt_PT    zh_CN
    de_CH    en_AU    en_US    fr_CH    ja_JP    nl_NL    ru_RU    zh_TW
    de_DE    en_CA    es_ES    fr_FR    ko_KR    pt_BR    sv_SE

The Apache Flex SDK has only been tested for the en_US locale. However, you are
unlikely to run into any issues with a SDK compiled for another locale.


Adobe Flash Player Version Support
==================================

The Apache Flex SDK defaults to using the Adobe Flash Player 11.1. The SDK can
be used with Flash Player versions 10.2 and up.

It is recommended that you update to the latest version of Adobe Flash Player.
Newer versions of the Adobe Flash player address security vulnerabilities, fix
bugs/issues, increase stability, and support new features.

To use the SDK for another version of the Adobe Flash Player either:

Change the flex-config.xml configuration file to specify another version of the
Adobe Flash Player.

OR

Change the playerglobal.version in the build.properties file to have a value
other than '11.1'. For this change to take effect, the SDK needs to be
recompiled. For example, to compile against the latest version of the Adobe
Flash Player, set the value of playerglobal.version to be '20.0' like so:

    playerglobal.version = 20.0

OR

Compile the SDK with a -Dplayerglobal.version=<version> option, where <version>
is the Adobe Flash Player version to compile the SDK against.

Different versions of the Adobe Flash Player require different versions of
playerglobal.swc.

These can be found at:

    http://download.macromedia.com/get/flashplayer/installers/archive/playerglobal/playerglobal10_2.swc
    http://download.macromedia.com/get/flashplayer/installers/archive/playerglobal/playerglobal10_3.swc

and

    http://download.macromedia.com/get/flashplayer/updaters/<version.major>/playerglobal<version.major>_<version.minor>.swc

(e.g. http://download.macromedia.com/get/flashplayer/updaters/11/playerglobal11_1.swc)

Copy the target playerglobal.swc to the directory:

    frameworks/libs/player/<version.major>.<version.minor>/playerglobal.swc
    
On OSX, running the script:

    ./ide/checkAllPlayerGlobals.sh 
 
will download any new or missing player global swcs.


Software Dependencies
=====================

The Apache Flex SDK uses third-party code that will be downloaded as part of
the build. In addition, you can choose to download optional third-party code
if you would like to take advantage of the features offered, and you agree to
the license terms.

The following dependencies have licenses which are, or are compatible with, the
Apache Version 2.0 license. You will not be prompted to acknowledge the
download. Most of the jars are installed in lib/external.

    batik - http://archive.apache.org/dist/xmlgraphics/batik/batik-1.6.zip
    commons-collections - http://archive.apache.org/dist/commons/collections/binaries/commons-collections-3.0.tar.gz
    commons-discovery - http://archive.apache.org/dist/commons/discovery/binaries/commons-discovery-0.2.tar.gz
    commons-logging - http://archive.apache.org/dist/commons/logging/binaries/commons-logging-1.0.4.tar.gz
    javacc - http://java.net/projects/javacc/downloads/download/javacc-5.0.tar.gz
    velocity - http://archive.apache.org/dist/velocity/engine/1.4/velocity-1.4.tar.gz
    xalan - http://archive.apache.org/dist/xml/xalan-j/binaries/xalan-j_2_6_0-bin.tar.gz
    xerces - http://archive.apache.org/dist/xerces/j/Xerces-J-bin.2.9.1.zip

The following dependencies have licenses which Apache considers to be
reciprocal licenses. These files are installed in lib/external
frameworks/libs respectively.

    saxon9 - http://sourceforge.net/projects/saxon/files/Saxon-B/9.1.0.8/saxonb9-1-0-8j.zip/download

RELEASE_NOTES
Apache Flex 4.16.1
==================

AIR and Flash Player Support
----------------------------
- Support Flash Player 21.0, 22.0, 23.0, 24.0, 25.0, 26.0 and 27.0
- Support for AIR 21.0, 22.0, 23.0, 24.0, 25.0, 26.0 and 27.0

Bug Fixes
---------
FLEX-35364 Flex installer.xml is failing occasionally
FLEX-35329 Initializing member variable/constant that is [Deprecated]
            incorrectly throws extra warning
FLEX-35321 LayoutManager initializes components which are no longer on stage
FLEX-35267 Spark DataGrid's GridHeaderViewLayout.getHeaderIndexAt() returns wrong header index when leftPadding is comparable to a column's width
FLEX-27509 AdvancedListBase itemMaskFreeList is null 

Known Issues
------------

Adobe Flash Builder Integration

Adobe Flash Builder 4.7 and Adobe Flash Builder 4.6 both contain a bug where,
when generating the .mxml file for a new project that uses Apache Flex SDK
4.10.0 or later Spark Application, Flash Builder will incorrectly insert the
attribute:

    layout="absolute"

This results in a compile error in the new project.  The remedy is to simply
remove the errant attribute.

The following wiki page has been set up to contain the latest news on Adobe's
attempt to correct this problem:

    https://cwiki.apache.org/confluence/display/FLEX/Adobe+Flash+Builder+'New+Project'+Bug

Adobe has provided a patch for Flash Builder 4.7 that resolves this issue:

    http://helpx.adobe.com/flash-builder/kb/flex-new-project-issue--.html


Apache Flex 4.16.0
==================

New Features
____________
FLEX-35122  Update to Font Awesome 4.7

AIR and Flash Player Support
----------------------------
- Support Flash Player 21.0, 22.0, 23.0 and 24.0
- Support for AIR 21.0, 22.0, 23.0 and 24.0

Bug Fixes
---------
FLEX-35165  Remove bundling of Apache 1.0 licensed software
FLEX-35126  DropDownController inside a pop up still listens to MOUSE_DOWN and
		MOUSE_DOWN_SOMEWHERE after the popup has closed
FLEX-35125  DropDownList inside a PopUpAnchor doesn't close when another DropDownList
		is selected
FLEX-35124  RTE 1009 in DropDownController.systemManager_mouseDownHandler() when inside
		a PopUpAnchor and other item is removed on click
FLEX-35123  DOM XSS vulnerability in history.js
FLEX-35109  falcon compiler's build.xml error
FLEX-35106  Incorrect date in datefield popup for fr_FR locale
FLEX-35100  TextFlowLine.recreateTextLine() throws a null object reference error
FLEX-35099  BusyIndicator has the wrong wrong rotation centre for iOS
FLEX-35097  How to compile AIR beta 22 with Flex SDK?
FLEX-35084  Ability to specify FileAttributes at compile time (mxmlc)
FLEX-35082  moveIndexFindRow doesn't accept integer values
FLEX-35078  Fatal when SelectionManager tries to reapply selection onto changed textFlow
FLEX-35056  Pressing escape on s:Alert causes a RangeError when Alert.CANCEL is passed
		as defaultButtonFlag
FLEX-35046  Broken icon using FlatSpark ButtonIcon on mobile apps
FLEX-35045  Flatspark ButtonIcon - icon does not display
FLEX-35043  Unexpected behaviour of ListCollectionView.itemUpdated() when only one
		parameter supplied
FLEX-35042  Mistake in example for Pattern 'O' in ASDoc for DateFormatter.formatString
FLEX-35039  ListCollectionView.setItemAt places the item in the wrong position if there
		are duplicate items in the collection
FLEX-35037  Add ArrayCollection unit tests donated by Justin
FLEX-35031  HierarchicalCollectionViewCursor cannot find sealed class instance
FLEX-35028  Error in DataGrid getHeaderRendererAt(-1)
FLEX-35025  Validator results cannot be intercepted if the value is null
FLEX-34935  Copy From Word Document And Pasting Into TextRegion Not Working Properly
FLEX-34926  Misplaced [PercentProxy] metatag in GridColumn
FLEX-34738  Not delete BusyIndicator
FLEX-34811  Email validator show allow apostrophe in email address
FLEX-34698  Flatspark ButtonIcon: Introduce 'iconFontPlacement' style
FLEX-34391  ScrollableStageText with large y value crashes app
FLEX-34088  CalloutButton closes when selecting an item in a spark Dropdown
FLEX-33646  The "required" property on Validator should be a getter/setter
FLEX-33311  Null pointer fatal in ConstraintLayout when executing Transition on
		DropDownList
FLEX-33058  RTE in HierarchicalCollectionViewCursor when a refresh occurs and old cursor
		had an enumerable property that new cursor item doesn't
FLEX-31948  When all the columns are invisible, pressing any character Key throws RTE
FLEX-27509  AdvancedListBase itemMaskFreeList is null
FLEX-26808  s:DataGrid Drag and Drop of multiple items doesn't work
FLEX-20313  SelectItem on AdvancedDataGrid use weird code


Known Issues
------------

Adobe Flash Builder Integration

Adobe Flash Builder 4.7 and Adobe Flash Builder 4.6 both contain a bug where,
when generating the .mxml file for a new project that uses Apache Flex SDK
4.10.0 or later Spark Application, Flash Builder will incorrectly insert the
attribute:

    layout="absolute"

This results in a compile error in the new project.  The remedy is to simply
remove the errant attribute.

The following wiki page has been set up to contain the latest news on Adobe's
attempt to correct this problem:

    https://cwiki.apache.org/confluence/display/FLEX/Adobe+Flash+Builder+'New+Project'+Bug

Adobe has provided a patch for Flash Builder 4.7 that resolves this issue:

    http://helpx.adobe.com/flash-builder/kb/flex-new-project-issue--.html



Apache Flex 4.15.0
==================

New Features
------------
FLEX-34837    Full support for sorting on complex fields of DataGrid rows
FLEX-34852    Sorting by complex fields in ListCollectionView
FLEX-34927    Allow the setting of an arbitrary property on an object via
                ObjectUtil
FLEX-34941    Allow custom validators to use the triggerEvent
FLEX-34945    Validators now able to validate nested properties

AIR and Flash Player Support
----------------------------
- Support Flash Player 18.0, 19.0 and 20.0
- Support for AIR 18.0, 19.0 and 20.0

Bug Fixes

ide/addAIRtoSDK.sh
#!/bin/sh -e

################################################################################
##
##  Licensed to the Apache Software Foundation (ASF) under one or more
##  contributor license agreements.  See the NOTICE file distributed with
##  this work for additional information regarding copyright ownership.
##  The ASF licenses this file to You under the Apache License, Version 2.0
##  (the "License"); you may not use this file except in compliance with
##  the License.  You may obtain a copy of the License at
##
##      http://www.apache.org/licenses/LICENSE-2.0
##
##  Unless required by applicable law or agreed to in writing, software
##  distributed under the License is distributed on an "AS IS" BASIS,
##  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##  See the License for the specific language governing permissions and
##  limitations under the License.
##
################################################################################

#
# Usage: addAIRtoSDK AIRversion "Apache Flex directory"
#
# This script will download and add the specified version of AIR to an SDK and
# update it's configuration.

# Process the parameters.

AIR_VERSION="$1"
OS=`uname`

if [[ "${AIR_VERSION}" != "24.0"
  && "${AIR_VERSION}" != "23.0"  && "${AIR_VERSION}" != "22.0"  && "${AIR_VERSION}" != "21.0"
  &&  "${AIR_VERSION}" != "20.0"  && "${AIR_VERSION}" != "19.0"  && "${AIR_VERSION}" != "18.0"
  && "${AIR_VERSION}" != "17.0" && "${AIR_VERSION}" != "16.0" && "${AIR_VERSION}" != "15.0" 
  && "${AIR_VERSION}" != "14.0" && "${AIR_VERSION}" != "13.0" && "${AIR_VERSION}" != "4.0" 
  && "${AIR_VERSION}" != "3.9" && "${AIR_VERSION}" != "3.8" && "${AIR_VERSION}" != "3.7" 
  && "${AIR_VERSION}" != "3.6" && "${AIR_VERSION}" != "3.5" && "${AIR_VERSION}" != "3.4" 
  && "${AIR_VERSION}" != "3.3" && "${AIR_VERSION}" != "3.2" && "${AIR_VERSION}" != "3.1" 
  && "${AIR_VERSION}" != "3.0" && "${AIR_VERSION}" != "2.7" && "${AIR_VERSION}" != "2.6" ]]
then
	echo Unknown version ${AIR_VERISON} of AIR. Versions 2.6, 2.7, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0,  23.0 and 24.0 are supported.
	exit 1;
fi

if [[ "${OS}" != "Darwin" && "${AIR_VERSION}" != "2.6" ]]
then
	echo "Only AIR version 2.6 is supported on Linux"
	exit 1;
fi

IDE_SDK_DIR="$2"

if [ "${IDE_SDK_DIR}" = "" ]
then
    echo Usage: $0 "AIR version" "Apache Flex directory"
    exit 1;
fi

echo The Apache Flex directory for the IDE is "$IDE_SDK_DIR"

# If this is an Apache Flex dir then there should be a NOTICE file.

if [ ! -f "${IDE_SDK_DIR}/NOTICE" ]
then
    echo ${IDE_SDK_DIR} does not appear to be an Apache Flex distribution.
    exit 1;
fi

agreeLicense()
{
	echo
	echo "Adobe AIR is licensed under the the Adobe AIR end user license agreement (EULA)."
	echo
	echo The Adobe AIR 3 EULA is specified here:
	echo http://www.adobe.com/products/eulas/pdfs/PlatformClients_PC_WWEULA-MULTI-20110809_1357.pdf
	echo
	echo In addition to the Adobe EULA license terms, you also agree to be bound by the third-party
	echo terms specified here:
	echo http://www.adobe.com/products/eula/third_party/
	echo
	echo Adobe recommends that you review all licensing terms.
	echo
	echo "Please type Y to agree to terms of the license >"
	
	read -n 1 accept
	echo
		
	if [[ "${accept}" = "Y" || "${accept}" = "y" ]]
	then
	   echo License accepted
	else
		exit 1;
	fi
}


downloadAIR()
{  	version=$1
   	airTempDir="${IDE_SDK_DIR}/frameworks/temp"
	mkdir -p "${airTempDir}"

    if [[ "${OS}" == "Darwin" ]]
    then
        airDownload="https://airdownload.adobe.com/air/mac/download/${version}/AdobeAIRSDK.tbz2"
    else
        airDownload="https://airdownload.adobe.com/air/lin/download/${version}/AdobeAIRSDK.tbz2"
    fi

    if [[ "${AIR_VERSION}" == "24.0" ]]
    then
        airDownload="https://airdownload.adobe.com/air/mac/download/24.0/AdobeAIRSDK.dmg"
	echo Downloading AIR ${version}
	echo from ${airDownload}
	curl ${airDownload} > "${airTempDir}/air.dmg"
	
	echo Extracting into SDK 
	hdiutil attach -nobrowse "${airTempDir}"/air.dmg
	cp -fR "/Volumes/AIR SDK/" "${IDE_SDK_DIR}"
	umount "/Volumes/AIR SDK"
    else
	echo Downloading AIR ${version}
	echo from ${airDownload}
	curl ${airDownload} > "${airTempDir}/air.tbz2"
	
	echo Extracting into SDK  
	tar xf "${airTempDir}/air.tbz2" -C "${IDE_SDK_DIR}"
    fi

    rm -rf "${airTempDir}"	
}

agreeLicense

downloadAIR ${AIR_VERSION}

# update config file

updatePlayerVersion() {
	playerversion=$1
	configFile=$2
	sed "s/<target-player>[0-9][0-9]\.[0-9]<\/target-player>/<target-player>${playerversion}<\/target-player>/" < "${configFile}"  > "${configFile}.tmp"
	mv "${configFile}.tmp" "${configFile}"
}

updateSWFVersion() {
	swfversion=$1
	configFile=$2
	sed "s/<swf-version>[0-9][0-9]<\/swf-version>/<swf-version>${swfversion}<\/swf-version>/" < "${configFile}"  > "${configFile}.tmp"
	mv "${configFile}.tmp" "${configFile}"
}

updatePlayerDescription() {
	airversion=$1
	configFile=$2
	
	echo Updating ${configFile}
	
	sed "s/AIR[0-9]\.[0-9]/AIR${airversion}/" < "${configFile}"  > "${configFile}.tmp"
	mv "${configFile}.tmp" "${configFile}"
}

configFiles=(
"${IDE_SDK_DIR}/frameworks/flex-config.xml"
"${IDE_SDK_DIR}/frameworks/air-config.xml"
"${IDE_SDK_DIR}/frameworks/airmobile-config.xml"
)

updatePlayerDescription "${AIR_VERSION}" "${IDE_SDK_DIR}/flex-sdk-description.xml"

for configFile in "${configFiles[@]}"
do
	echo Updating ${configFile}
	
	# 24.0 needs FP 24 and swf version 35
	if [ ${AIR_VERSION} = "23.0" ]
	then
		updatePlayerVersion 24.0 "${configFile}"
		updateSWFVersion 35 "${configFile}"
	fi

	# 23.0 needs FP 23 and swf version 34
	if [ ${AIR_VERSION} = "23.0" ]
	then
		updatePlayerVersion 23.0 "${configFile}"
		updateSWFVersion 34 "${configFile}"
	fi

	# 22.0 needs FP 22 and swf version 33
	if [ ${AIR_VERSION} = "22.0" ]
	then
		updatePlayerVersion 22.0 "${configFile}"
		updateSWFVersion 33 "${configFile}"
	fi

	# 21.0 needs FP 21 and swf version 32
	if [ ${AIR_VERSION} = "21.0" ]
	then
		updatePlayerVersion 21.0 "${configFile}"
		updateSWFVersion 32 "${configFile}"
	fi

	# 20.0 needs FP 20 and swf version 31
	if [ ${AIR_VERSION} = "20.0" ]
	then
		updatePlayerVersion 20.0 "${configFile}"
		updateSWFVersion 31 "${configFile}"
	fi
	
	# 19.0 needs FP 19 and swf version 30
	if [ ${AIR_VERSION} = "19.0" ]
	then
		updatePlayerVersion 19.0 "${configFile}"
		updateSWFVersion 30 "${configFile}"
	fi

	# 18.0 needs FP 18 and swf version 29
	if [ ${AIR_VERSION} = "18.0" ]
	then
		updatePlayerVersion 18.0 "${configFile}"
		updateSWFVersion 29 "${configFile}"
	fi	
	
	# 17.0 needs FP 17 and swf version 28
	if [ ${AIR_VERSION} = "17.0" ]
	then
		updatePlayerVersion 17.0 "${configFile}"
		updateSWFVersion 28 "${configFile}"
	fi	
	
	# 16.0 needs FP 16 and swf version 27
	if [ ${AIR_VERSION} = "16.0" ]
	then
		updatePlayerVersion 16.0 "${configFile}"
		updateSWFVersion 27 "${configFile}"
	fi	
	
	# 15.0 needs FP 15 and swf version 26
	if [ ${AIR_VERSION} = "15.0" ]
	then
		updatePlayerVersion 15.0 "${configFile}"
		updateSWFVersion 26 "${configFile}"
	fi	

	# 14.0 needs FP 14 and swf version 25
	if [ ${AIR_VERSION} = "14.0" ]
	then
		updatePlayerVersion 14.0 "${configFile}"
		updateSWFVersion 25 "${configFile}"
	fi	
	
	# 13.0 needs FP 13 and swf version 24
	if [ ${AIR_VERSION} = "13.0" ]
	then
		updatePlayerVersion 13.0 "${configFile}"
		updateSWFVersion 24 "${configFile}"
	fi	
	
	# 4.0 needs FP 12 and swf version 23
	if [ ${AIR_VERSION} = "4.0" ]
	then
		updatePlayerVersion 12.0 "${configFile}"
		updateSWFVersion 23 "${configFile}"
	fi	
		
	# 3.8 needs FP 11.9 and swf version 22
	if [ ${AIR_VERSION} = "3.9" ]
	then
		updatePlayerVersion 11.9 "${configFile}"
		updateSWFVersion 22 "${configFile}"
	fi	
		
	# 3.8 needs FP 11.8 and swf version 21
	if [ ${AIR_VERSION} = "3.8" ]
	then
		updatePlayerVersion 11.8 "${configFile}"
		updateSWFVersion 21 "${configFile}"
	fi	

	# 3.7 needs FP 11.7 and swf version 20
	if [ ${AIR_VERSION} = "3.7" ]
	then
		updatePlayerVersion 11.7 "${configFile}"
		updateSWFVersion 20 "${configFile}"
	fi	
	
	# 3.6 needs FP 11.6 and swf version 19
	if [ ${AIR_VERSION} = "3.6" ]
	then
		updatePlayerVersion 11.6 "${configFile}"
		updateSWFVersion 19 "${configFile}"
	fi

	# 3.5 needs FP 11.5 and swf version 18
	if [ ${AIR_VERSION} = "3.5" ]
	then
		updatePlayerVersion 11.5 "${configFile}"
		updateSWFVersion 18 "${configFile}"
	fi
	
	# 3.4 needs FP 11.4 and swf version 17
	if [ ${AIR_VERSION} = "3.4" ]
	then
		updatePlayerVersion 11.4 "${configFile}"
		updateSWFVersion 17 "${configFile}"
	fi
	
	# 3.3 needs FP 11.3 and swf version 16
	if [ ${AIR_VERSION} = "3.3" ]
	then
		updatePlayerVersion 11.3 "${configFile}"
		updateSWFVersion 16 "${configFile}"
	fi
		
	# 3.2 needs FP 11.2 and swf version 15
	if [ ${AIR_VERSION} = "3.2" ]
	then
		updatePlayerVersion 11.2 "${configFile}"
		updateSWFVersion 15 "${configFile}"
	fi
	
	# 3.1 needs FP 11.1 and swf version 14
	if [ ${AIR_VERSION} = "3.1" ]
	then
		updatePlayerVersion 11.1 "${configFile}"
		updateSWFVersion 14 "${configFile}"
	fi
	
	# 3.0 needs FP 11.0 and swf version 13
	if [ ${AIR_VERSION} = "3.0" ]
	then
		updatePlayerVersion 11.0 "${configFile}"
		updateSWFVersion 13 "${configFile}"
	fi
	
	# 2.7 needs FP 10.3 and swf version 12
	if [ ${AIR_VERSION} = "2.7" ]
	then
		updatePlayerVersion 10.3 "${configFile}"
		updateSWFVersion 12 "${configFile}"
	fi
	
	# 2.6 needs FP 10.2 and swf version 11
	if [ ${AIR_VERSION} = "2.6" ]
	then
		updatePlayerVersion 10.2 "${configFile}"
		updateSWFVersion 11 "${configFile}"
	fi
done


